app/controllers/application_controller.rb

```diff
@@ -424,7 +424,7 @@ def check_ownership_or_admin
   # Make sure the user is allowed to see a specific page
   # includes a special case for restricted works and series, since we want to encourage people to sign up to read them
   def check_visibility
-    if @check_visibility_of.respond_to?(:restricted) && @check_visibility_of.restricted && User.current_user.nil?
+    if @check_visibility_of.respond_to?(:restricted) && @check_visibility_of.restricted && !(current_user || current_admin)
```

Comment:
`User.current_user` is assigned here:
https://github.com/otwcode/otwarchive/blob/0da96bee3ca1587403ec5b73e8d3c2d192e7b0c6/app/controllers/application_controller.rb#L301-L303
It's equivalent to the code you've replaced it with, as far as I can tell.

nginx is configured to hide the user's cookies for downloads, I believe, so all visitors are being treated as logged out by the DownloadsController. Which means that the only way to make restricted works visible to logged-in users is to make them visible to everyone, which isn't quite compatible with the current definition of check_visibility. (The old code called `check_visibility` too, but never ran into this issue because `@check_visibility_of` was set only after `check_visibility` ran, making it a no-op.)

If you want to keep the `before_action :check_visibility` filter, you could try modifying the behavior when the `controller_name` is "downloads." Though it's probably cleaner to just write a custom visibility check for the DownloadsController.

(The other alternative involves playing around with nginx.)

Reply:
It is indeed -- that's actually where I got it from. It seemed like `User.current_user` wasn't set, so I thought maybe I could get around it by checking `current_user` and `current_admin` directly.

I'm going to ping you elsewhere re: the nginx option, if you don't mind.